src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// We only build the Source Generator in the netstandard target
+#if NETSTANDARD
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace CSharpSyntaxGenerator
+{
+    public abstract class CachingSourceGenerator : ISourceGenerator
+    {
+        private static Tuple<ImmutableArray<byte>, ImmutableArray<(string hintName, SourceText sourceText)>>? s_cachedResult;
+
+        protected abstract bool TryGetRelevantInput(in GeneratorExecutionContext context, out AdditionalText? input, out SourceText? inputText);
+
+        protected abstract bool TryGenerateSources(
+            AdditionalText input,
+            SourceText inputText,
+            out ImmutableArray<(string hintName, SourceText sourceText)> sources,
+            out ImmutableArray<Diagnostic> diagnostics,
+            CancellationToken cancellationToken);
+
+        public virtual void Initialize(GeneratorInitializationContext context)
+        {
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            if (!TryGetRelevantInput(in context, out var input, out var inputText))
+            {
+                return;
+            }
+
+            // Get the current input checksum, which will either be used for verifying the current cache or updating it
+            // with the new results.
+            var currentChecksum = inputText.GetChecksum();
+
+            // Read the current cached result once to avoid race conditions
+            if (s_cachedResult is { } cachedResult
+                && cachedResult.Item1.SequenceEqual(currentChecksum))
+            {
+                AddSources(in context, sources: cachedResult.Item2, currentChecksum, CacheBehavior.None);
```

Comment:
If the previous run generated diagnostics, it seems like we may lose them the next time around.

Reply:
If the previous run generated diagnostics, the result is not cached and the next run will (theoretically) produce the same diagnostics.

Reply:
ah ok :)